### ISSUE TITLE
Longhorn cannot work on k8s v1.11

### ISSUE THREAD

**Issue by niusmallnan** [rancher/longhorn#162]
Deploy k8s v1.11 by RKE  v0.1.9-rc6.

I can't create a volume, and replicas cannot be scheduled.
The disk information has not been updated to node.

```
{
	"data": [{
		"actions": {
			"diskUpdate": "…/v1/nodes/172.31.31.165?action=diskUpdate",
		},
		"address": "10.42.0.6",
		"allowScheduling": true,
		"disks": {
			"26f5250955ccb740": {
				"State": "",
				"allowScheduling": true,
				"path": "/var/lib/rancher/longhorn/",
				"storageAvailable": 0,
				"storageMaximum": 30123499520,
				"storageReserved": 0,
				"storageScheduled": 0,
			},
		},
		"id": "172.31.31.165",
		"links": {
			"self": "…/v1/nodes/172.31.31.165",
		},
		"mountPropagation": false,
		"name": "172.31.31.165",
		"state": "up",
		"type": "node",
	}, ],
	"links": {
		"self": "…/v1/nodes",
	},
	"resourceType": "node",
	"type": "collection",
}
```

Maybe we need to update the client-go version and re-adapt the CRD controllers.

**Comment by yasker** [comment]
Anything from manager's log?

**Comment by yasker** [comment]
Storage available is 0 on the disk, but maximum seems correct. @JacieChao Can you check this?

**Comment by JacieChao** [comment]
Sure. I will check it.

**Comment by niusmallnan** [comment]
The  manager's log:
```
time="2018-08-03T07:41:41Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:42Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:42Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:42Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:43Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:43Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:43Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:43Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:44Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:44Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:44Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:45Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:45Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:45Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:46Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:46Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:46Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:46Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
time="2018-08-03T07:41:47Z" level=debug msg="Requeue 172.31.31.165 due to conflict"
```
It looks like it is in an infinite loop.

**Comment by yasker** [comment]
There are multiple issues here:
1. Our node controller didn't carry an always up to date (to this node's knowledge) reference to the node object when running the reconciliation loop. There are three places that the node object can be updated. Once one of them got updated in one place, all the remaining ones won't be able to be updated because the reference to the object wasn't updated to the latest version. That caused the majority of conflicts.
2. But this issue wasn't exposed in the previous version of Kubernetes, because seems Kubernetes API server will compare the object that needs to be updated to a previous version (should be in the server side since we didn't update the client-go), and only update it when change happens. This behavior eliminates many conflicts due to write the same object to the API server. But due to some reason, I observed that in v1.11 this is no longer the behavior. API server will accept the update and assign a new revision number to it.
3. Combined item 1 and item 2, the end result is there is no way for the node controller to update the node object. Because the first update inside the node controller is always the `up/down` state sync. This was done without checking if the previous state is the same. So once this action succeeds (and it will always succeed for the current node), all the following update will be blocked by the revision conflict. That's why the disk status cannot be updated, so no data for available storage, then no replica can be scheduled.

The fixes are two parts, the first one will put all the update to the node into one place, so no more revision change during the reconciliation loop. The second one will add the check for the identical object before updating the object.

Other than that, we should change node up/down state into a condition as well. The condition has a LastProbeTimestamp, and LastTransitionTime which can be used to determine when this node has been checked and state changed, which can be useful to user.
